runtime/browser/xwalk_content_base.cc

```diff
@@ -0,0 +1,138 @@
+    : web_contents_(web_contents.Pass()) {
+#if defined(OS_ANDROID)
+  content::WebContents* contents = web_contents_.get();
+  XWalkAutofillClient* autofill_manager_delegate =
+      XWalkAutofillClientAndroid::FromWebContents(contents);
+  if (autofill_manager_delegate)
+    InitAutofillIfNecessary(autofill_manager_delegate->GetSaveFormData());
+#else
+  if (!XWalkBrowserContext::GetDefault()->SaveFormDataEnabled())
+    return;
+  InitAutofillIfNecessary(true);
+  SetXWalkUserData();
+#endif
+
+  CreateUserPrefServiceIfNecessary();
+  PrefService* pref_service =
+    user_prefs::UserPrefs::Get(XWalkBrowserContext::GetDefault());
+  pref_change_registrar_.Init(pref_service);
+  if (pref_service) {
+    base::Closure renderer_callback = base::Bind(
+      &XWalkContentBase::UpdateRendererPreferences, base::Unretained(this));
+    pref_change_registrar_.Add("intl.accept_languages", renderer_callback);
+  }
+}
+
+void XWalkContentBase::UpdateRendererPreferences() {
+  content::RendererPreferences* prefs =
+    web_contents_->GetMutableRendererPrefs();
+  PrefService* pref_service =
+    user_prefs::UserPrefs::Get(XWalkBrowserContext::GetDefault());
+  const std::string accept_languages =
+    pref_service->GetString("intl.accept_languages");
+  prefs->accept_languages = accept_languages;
+  web_contents_->GetRenderViewHost()->SyncRendererPrefs();
+  XWalkBrowserContext* browser_context =
+    XWalkRunner::GetInstance()->browser_context();
+  CHECK(browser_context);
+  browser_context->UpdateAcceptLanguages(accept_languages);
+}
+
+void XWalkContentBase::InitAutofillIfNecessary(bool enabled) {
+  // Do not initialize if the feature is not enabled.
+  if (!enabled)
+    return;
+  // Check if the autofill driver factory already exists.
+  content::WebContents* web_contents = web_contents_.get();
+  if (autofill::ContentAutofillDriverFactory::FromWebContents(web_contents))
+    return;
+
+  CreateUserPrefServiceIfNecessary();
+  std::string locale;
+#if defined (OS_ANDROID)
+  XWalkAutofillClientAndroid::CreateForWebContents(web_contents);
+  autofill::ContentAutofillDriverFactory::CreateForWebContentsAndDelegate(
+    web_contents,
+    XWalkAutofillClientAndroid::FromWebContents(web_contents),
+    base::android::GetDefaultLocale(),
+    autofill::AutofillManager::DISABLE_AUTOFILL_DOWNLOAD_MANAGER);
+#else
+  XWalkAutofillClientDesktop::CreateForWebContents(web_contents);
+  autofill::ContentAutofillDriverFactory::CreateForWebContentsAndDelegate(
+    web_contents,
+    XWalkAutofillClientDesktop::FromWebContents(web_contents),
+    XWalkContentBrowserClient::Get()->GetApplicationLocale(),
+    autofill::AutofillManager::DISABLE_AUTOFILL_DOWNLOAD_MANAGER);
+#endif
+}
+
+void XWalkContentBase::CreateUserPrefServiceIfNecessary() {
+  XWalkBrowserContext* browser_context =
+    XWalkBrowserContext::FromWebContents(web_contents_.get());
+  browser_context->CreateUserPrefServiceIfNecessary();
+}
+
+void XWalkContentBase::SetXWalkUserData() {
```

Comment:
calling directly web_contents_->SetUserData(.. does not hurt, so maybe this method is not needed?